drivers/FlashIAP.cpp

```diff
@@ -86,21 +94,39 @@ int FlashIAP::program(const void *buffer, uint32_t addr, uint32_t size)
 {
     uint32_t page_size = get_page_size();
     uint32_t current_sector_size = flash_get_sector_size(&_flash, addr);
-    // addr and size should be aligned to page size, and multiple of page size
-    // page program should not cross sector boundaries
-    if (!is_aligned(addr, page_size) ||
-        !is_aligned(size, page_size) ||
-        (size < page_size) ||
-        (((addr % current_sector_size) + size) > current_sector_size)) {
+    uint32_t chunk, prog_size;
+    const uint8_t *buf = (uint8_t *) buffer;
+
+    // addr should be aligned to page size
+    if (!is_aligned(addr, page_size) || (!buffer)) {
         return -1;
     }
 
     int ret = 0;
     _mutex->lock();
-    if (flash_program_page(&_flash, addr, (const uint8_t *)buffer, size)) {
-        ret = -1;
+    while (size) {
```

Comment:
while reviewing this, this is similar to https://github.com/ARMmbed/mbed-os/issues/6077 (this is however only program, but erase should do the same) ?

Reply:
Not sure I understand the comment. Erase seems to be following a similar algorithm as program now. What's the problem?

Reply:
Oh I see the problem now.
My thought here is that the entire check should just be eliminated. Why not erase everything that's in the range, and let the driver worry about alignments, not the user.

Reply:
So if I understand correctly, an unaligned erase does this?

```
            --- sector 0 ---  ->  --- sector 0 ---    
            data                  data
            data                  data
            data                  data
            --- sector 1 ---      --- sector 1 ---
            data                  ff
          > data                  ff
unaligned | data                  ff
erase     | --- sector 2 ---      --- sector 2 ---
          > data                  ff
            data                  ff
            data                  ff
```

That seems like the sort of thing we'd want to error on.

Reply:
So if I understand correctly, an unaligned erase does this?

```
            --- sector 0 ---  ->  --- sector 0 ---    
            data                  data
            data                  data
            data                  data
            --- sector 1 ---      --- sector 1 ---
            data                  ff
          > data                  ff
unaligned | data                  ff
erase     | --- sector 2 ---      --- sector 2 ---
          > data                  ff
            data                  ff
            data                  ff
```

That seems like the sort of thing we'd want to error on.

Reply:
As said, don't think there should even be an unaligned erase check. Don't even know what it stands for.

Reply:
Let me rephrase: I believe the only acceptable validation here is to check that the address is on a sector boundary, and that the size is a sum of all the consecutive sector sizes. The only way to achieve this is by iterating on the sectors for the validation, and then iterating again for the erase action. It's either that or check nothing.

Reply:
Ah ok. Don't let me block this pr then.

Could we check that address + size is on a sector boundary? Would that be cheap?

Reply:
Yeah, figured this out after writing the answer. This would be the correct check.